### PR TITLE
[BEAM-12313] - Initial changes to support Java 17 SDK Harness docker image

### PR DESCRIPTION
**Pull request by fernando-wizeline**
First "draft" to review the changes required to publish the Java 17 SDK Harness docker image.

@emilymye, if I'm not mistaken, you were in charge of the same effort but for the Java 11 SDK. Can you help me to review this and let me know what I'm missing?
Thanks a lot for the help!
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by kileys** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by kileys** [comment]
Run Python PreCommit

**Comment by kileys** [comment]
@kennknowles can you take a look?

File: sdks/java/testing/test-utils/src/test/java/org/apache/beam/sdk/testutils/jvmverification/JvmVerification.java [hunk 1]
```diff
@@ -55,6 +54,11 @@ public void verifyTestCodeIsCompiledWithJava11() throws IOException {
     assertEquals(v11, getByteCodeVersion(JvmVerification.class));
   }
 
+  @Test
+  public void verifyTestCodeIsCompiledWithJava17() throws IOException {
+    assertEquals(v17, getByteCodeVersion(JvmVerification.class));
```
Comment (kennknowles) [inline sdks/java/testing/test-utils/src/test/java/org/apache/beam/sdk/testutils/jvmverification/JvmVerification.java:59]: Need to add an entry to `versionMapping`, right?

**Review by kennknowles — CHANGES_REQUESTED** [review]
Seems OK. We probably need the container to be checked in first before the Jenkins jobs can succeed.

**Comment by fernando-wizeline** [comment]
> Seems OK. We probably need the container to be checked in first before the Jenkins jobs can succeed.

Thanks @kennknowles! Does Infra takes care of checking in the container? I need to open a ticket in JIRA for that, right?

**Comment by kennknowles** [comment]
> > Seems OK. We probably need the container to be checked in first before the Jenkins jobs can succeed.
> 
> Thanks @kennknowles! Does Infra takes care of checking in the container? I need to open a ticket in JIRA for that, right?

Our nightly snapshot pushes containers. I'm not super familiar with it - not sure if your change will already cause that to happen or not.

**Comment by pabloem** [comment]
https://ci-beam.apache.org/job/beam_SeedJob/8826/console

**Comment by pabloem** [comment]
seed job looks good : )

**Comment by fernando-wizeline** [comment]
Run Java PreCommit

**Comment by fernando-wizeline** [comment]
@pabloem what do you think? Would it be ok to merge this?

**Comment by fernando-wizeline** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by pabloem** [comment]
this looks fine to me, though I'll let @kennknowles finish the review as he had started doing it, and must have more context. Happy to help with seed jobs / etc

**Comment by fernando-wizeline** [comment]
> this looks fine to me, though I'll let @kennknowles finish the review as he had started doing it, and must have more context. Happy to help with seed jobs / etc

Thanks @pabloem!
Hi @kennknowles! It seems we need to open a ticket similar to this one https://issues.apache.org/jira/browse/INFRA-20866 to create a sdk17 image. I tried opening a new one but I don't have the right JIRA permissions. That and the changes already contained in this PR should take care of everything related to the SDK 17 Harness.

**Comment by kileys** [comment]
I had permissions to create a ticket: https://issues.apache.org/jira/browse/INFRA-22561

**Comment by fernando-wizeline** [comment]
> I had permissions to create a ticket: https://issues.apache.org/jira/browse/INFRA-22561

Hi @kileys!
The repo was created but now we need someone to push an image. Do you know who can help us with that? It looks like last time Emily helped with that but I'm not sure if she can help this time.
Thanks!

**Comment by kileys** [comment]
Do we just need to run the Jenkins job to publish an image?

**Comment by kileys** [comment]
@fernando-wizeline were all the tests run successfully after the seed job? I don't see any runs on Jenkins

**Comment by fernando-wizeline** [comment]
> @fernando-wizeline were all the tests run successfully after the seed job? I don't see any runs on Jenkins

hi @kileys! Yeah they did run successfully twenty days ago, but since then I added new ones.
@pabloem, can you help us by  kicking off the seed job again?

**Comment by kileys** [comment]
After the seed job, you'll have to use the trigger phrase to actually have the tests run. The seed job doesn't start the new tests e.g. https://ci-beam.apache.org/job/beam_LoadTests_Java_ParDo_Dataflow_V2_Streaming_Java17_PR/

**Comment by tvalentyn** [comment]
Seed job: https://ci-beam.apache.org/job/beam_SeedJob/8919/

**Comment by kennknowles** [comment]
Are you unblocked here?

**Comment by fernando-wizeline** [comment]
> Are you unblocked here?

Hi @kennknowles!
I am, but some of the new Java 17 tests failed during the execution of the seedjob.
It looks like a matter of setting up the Java Home correctly.
I'll be triaging that this week.

**Comment by fernando-wizeline** [comment]
Run Load Tests Java 11 CoGBK Dataflow V2 Streaming

File: buildSrc/build.gradle.kts [hunk 2]
```diff
@@ -35,29 +35,29 @@ repositories {
...
-  runtime("org.checkerframework:checkerframework-gradle-plugin:0.5.16")                                 // Enable enhanced static checking plugin
+  runtimeOnly("net.ltgt.gradle:gradle-apt-plugin:0.21")                                                    // Enable a Java annotation processor
+  runtimeOnly("com.google.protobuf:protobuf-gradle-plugin:0.8.13")                                          // Enable proto code generation
+  runtimeOnly("io.spring.gradle:propdeps-plugin:0.0.9.RELEASE")                                            // Enable provided and optional configurations
+  runtimeOnly("com.commercehub.gradle.plugin:gradle-avro-plugin:0.11.0")                                   // Enable Avro code generation
+  runtimeOnly("com.diffplug.spotless:spotless-plugin-gradle:5.6.1")                                       // Enable a code formatting plugin
+  runtimeOnly("gradle.plugin.com.github.blindpirate:gogradle:0.11.4")                                      // Enable Go code compilation
+  runtimeOnly("gradle.plugin.com.palantir.gradle.docker:gradle-docker:0.22.0")                             // Enable building Docker containers
```
Comment (tvalentyn) [inline buildSrc/build.gradle.kts]: do we need this one? i remember we switched to a different plugin earlier (the one that is used in other places in Beam).

**Comment by fernando-wizeline** [comment]
hi @tvalentyn !
Can you help me to run the seed job for this again?
Thanks!

**Comment by tvalentyn** [comment]
did that (it failed).

**Comment by fernando-wizeline** [comment]
> did that (it failed).

Thanks @tvalentyn!
Can you share the link to the results page, please?

**Comment by fernando-wizeline** [comment]
> > did that (it failed).
> 
> Thanks @tvalentyn! Can you share the link to the results page, please?

Found it, sorry!!!

**Comment by tvalentyn** [comment]
no worries, feel free to ping this when you need another run.

**Comment by fernando-wizeline** [comment]
hi @tvalentyn, 
Is there any chance of running the seed-job again?
Thanks!

**Comment by ibzib** [comment]
https://ci-beam.apache.org/job/beam_SeedJob/9042

**Comment by fernando-wizeline** [comment]
Run Load Tests Java 17 CoGBK Dataflow V2 Streaming

**Comment by fernando-wizeline** [comment]
Run Load Tests Java 11 CoGBK Dataflow V2 Streaming

**Comment by fernando-wizeline** [comment]
Run Direct ValidatesRunner Java 17

**Comment by fernando-wizeline** [comment]
Run Direct ValidatesRunner Java 11

**Comment by ibzib** [comment]
Looks like the seed job is permared: https://issues.apache.org/jira/browse/BEAM-13581

**Comment by fernando-wizeline** [comment]
Hi @lukecwik!
It looks like the CoGBK load tests rely on the Caches class which in turn rely on the jamm library.
The jamm library is using a the setAccessible method which, per JEP403, no longer works and throws an InaccessibleObjectException instead.
What would you say is the best approach in this case?
Thanks a lot for the help! Sorry if I’m not seeing a pretty obvious solution here.

**Comment by lukecwik** [comment]
--add-opens should be done for all modules that exist on the classpath and only access to jamm should be provided. If jamm doesn't support java 9 modules yet then we can modify the jar and embed a module-info in it so that we don't open all modules and packages to everyone.

**Comment by aaltay** [comment]
What is the next step on this PR?

**Comment by kileys** [comment]
This can be closed. There's a new PR with the jamm workaround

**Pull request closed without merging**